workers/unity/Assets/Fps/Scripts/WorldTiles/MapTemplate.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using Random = System.Random;
+
+namespace Fps
+{
+    [CreateAssetMenu(fileName = "MapTemplate", menuName = "Improbable/Map Template")]
+    public class MapTemplate : ScriptableObject
+    {
+        [SerializeField] internal Texture2D templateBitmap;
+        [SerializeField] internal TileTypeCollection[] tileCollections;
+        [SerializeField] internal TileTypeCollection defaultTileCollection;
+        [SerializeField] internal float unitSize = 1;
+
+        private readonly Dictionary<Color32, TileTypeCollection> tileLookup = new Dictionary<Color32, TileTypeCollection>();
+
+        public GameObject GetTileForLocation(Vector2 location, Random random)
+        {
+            TileTypeCollection collection = defaultTileCollection;
+
+            // Offset location so 0,0 is the middle of the bitmap
+            location += new Vector2(templateBitmap.width / 2.0f - 1, templateBitmap.height / 2.0f - 1);
+            location /= unitSize;
+            Vector2Int texelLocation = Vector2Int.FloorToInt(location);
+
+            if (texelLocation.x >= 0 && texelLocation.x < templateBitmap.width &&
+                texelLocation.y >= 0 && texelLocation.y < templateBitmap.height)
+            {
+                Color32 index = ColorToLookup(templateBitmap.GetPixel(texelLocation.x, texelLocation.y));
+                if (index.a == 0xff)
+                {
+                    if (!tileLookup.TryGetValue(index, out collection))
+                    {
+                        Debug.Log($"Unknown color: {index}\nThere are {tileLookup.Count} colors available.");
+                        PrintColors();
+                        collection = defaultTileCollection;
+                    }
+                }
+            }
+
+            return collection.GetRandomTile(random);
+        }
+
+        private void OnEnable()
+        {
+            InitializeLookup();
+        }
+
+        private void PrintColors()
+        {
+            var message = tileLookup.Select(element => $"[{element.Key}]")
+                .Aggregate("", (current, next) => $"{current} {next}");
+            Debug.Log(message);
+        }
+
+        private void OnValidate()
+        {
+            // Check for issues
+            if (!templateBitmap.isReadable)
+            {
+                Debug.LogError($"template bitmap {templateBitmap.name} must be Read/Write enabled.");
+            }
+
+            if (templateBitmap.format != TextureFormat.RGBA32)
+            {
+                Debug.LogError($"template bitmap {templateBitmap.name} must be RGBA32.");
+            }
+
+            if (defaultTileCollection == null)
+            {
+                Debug.LogError($"Default Tile Collection must be set.");
+            }
+
+            if (unitSize == 0)
```

Comment:
You could constrain the field with https://unity3d.com/learn/tutorials/topics/tips/show-public-variables-sliders-rangemin-max instead of auto changing it in validate?

Reply:
Also what happens if its negative

Reply:
Fixed the negative case. The slider is an issue since it would need a max value as well.

Reply:
Max value = max float 😛